src/transformers/models/albert/modeling_albert.py

```diff
@@ -249,10 +249,14 @@ def forward(
             else:
                 token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=self.position_ids.device)
```

Comment:
if we are going to switch it a moment later to another rank, as proposed next, this is wasteful.

We should get the device figured out first and then create the variable directly on the target device.

Reply:
Yes. we can add another one else statement for this. 
But let's start with the model without token type id as you said above.

Reply:
why adding an else statement, I don't quite follow?

I suggested we figure out the device first, and then in one go create the variable on the right device. i.e. your rank code goes before the creation, the correct device is set and then there is no `.to()` to do.

Reply:
I thought of following code.

```python
        if token_type_ids is None:
            if hasattr(self, "token_type_ids"):
                buffered_token_type_ids = self.token_type_ids[:, :seq_length]
                buffered_token_type_ids_expanded = buffered_token_type_ids.expand(input_shape[0], seq_length)
                token_type_ids = buffered_token_type_ids_expanded
            elif model_parallel:  # <-- new variable
                token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=os.getenv("LOCAL_RANK"))
            else:
                token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=self.position_ids.device)
```

Reply:
If model parallel is not applied, the LOCAL_RANK variable does not exist.

Reply:
thank you for clarifying with code - much easier to understand. I think that's a good way as you proposed. with a few suggestions:

1. we already have model_parallel - which is used for naive pipeline parallel

2. we are going to follow Deepspeed's naming conventions - so right now we are porting tensor parallelism (TP). As parallelfomers implements TP. Once we complete this project we can work on PP (pipeline parallelism).

Please refer to: https://huggingface.co/transformers/parallelism.html#concepts

Reply:
but still I'd prefer to abstract `os.getenv("LOCAL_RANK"))` and include meaningful exceptions should it be invalid for whatever reason. i.e. `transfromers` needs to have a defined API to get the rank and not rely just on env var.

Reply:
You are right. It would be good to implement mpu like nvidia megatron and provide it as a utility function.

Reply:
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/mpu/initialize.py#L255

Reply:
we will absolutely need to MPU anyway, so let's to do it right from the get-going!

You have a choice of Megatron or deepspeed MPU versions, I may have seen some other.

I think I may have even started to port one while trying to plug my PP PR into Deepspeed's 3D. yes, I did:
https://github.com/huggingface/transformers/pull/9765/files#diff-48e672da3865f77a2e1d38954e8e075c0f1e02c7306f163847b9b8ecc56ede24
I see I took it from megatron.

Reply:
Perhaps we can re-use some other parts of my work, but things have probably changed quite a lot in transformers to try to rescue much. there was some of the mpu use in the modeling_t5.py, but if I remember it wasn't quite finished.